### PR TITLE
HDDS-10741. Mark value objects as Immutable

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

Some value objects were already changed to be immutable as part of HDDS-10519.  Annotate them using `net.jcip.annotations.Immutable` to let SpotBugs catch any regressions (or false assumptions).

https://issues.apache.org/jira/browse/HDDS-10741

## How was this patch tested?

CI:
https://github.com/adoroszlai/ozone/actions/runs/8799185403/job/24147719729

Temporarily marked `OmKeyInfo` as `@Immutable`, verified that `findbugs` check reports violation:

```
M B JCIP: OmKeyInfo.dataSize should be final since org.apache.hadoop.ozone.om.helpers.OmKeyInfo is marked as Immutable.    In OmKeyInfo.java
M B JCIP: OmKeyInfo.encInfo should be final since org.apache.hadoop.ozone.om.helpers.OmKeyInfo is marked as Immutable.    In OmKeyInfo.java
M B JCIP: OmKeyInfo.fileName should be final since org.apache.hadoop.ozone.om.helpers.OmKeyInfo is marked as Immutable.    In OmKeyInfo.java
M B JCIP: OmKeyInfo.isFile should be final since org.apache.hadoop.ozone.om.helpers.OmKeyInfo is marked as Immutable.    In OmKeyInfo.java
M B JCIP: OmKeyInfo.keyLocationVersions should be final since org.apache.hadoop.ozone.om.helpers.OmKeyInfo is marked as Immutable.    In OmKeyInfo.java
M B JCIP: OmKeyInfo.keyName should be final since org.apache.hadoop.ozone.om.helpers.OmKeyInfo is marked as Immutable.    In OmKeyInfo.java
M B JCIP: OmKeyInfo.modificationTime should be final since org.apache.hadoop.ozone.om.helpers.OmKeyInfo is marked as Immutable.    In OmKeyInfo.java
M B JCIP: OmKeyInfo.replicationConfig should be final since org.apache.hadoop.ozone.om.helpers.OmKeyInfo is marked as Immutable.    In OmKeyInfo.java
```

### CODE REVIEW

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/VolumeArgs.java [hunk 1]
```diff
@@ -18,13 +18,15 @@
...
+import net.jcip.annotations.Immutable;
 
 import java.util.HashMap;
 import java.util.Map;
 
 /**
  * A class that encapsulates the createVolume Args.
  */
+@Immutable
 public final class VolumeArgs {
```
Comment (szetszwo) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/VolumeArgs.java:30]: This `VolumeArgs` class is not used anywhere.  Let's remove it instead?

**Review by szetszwo — COMMENTED** [review]
@adoroszlai , thanks for working on this!  Please see the comments inlined.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/VolumeArgs.java (hunk 1 above)
Comment (adoroszlai) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/VolumeArgs.java]: Looks like I had changed the wrong class.  Thanks for spotting it.

Removed this one, annotated the real one.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/BucketArgs.java [hunk 2]
```diff
@@ -35,6 +36,7 @@
  * This class encapsulates the arguments that are
  * required for creating a bucket.
  */
+@Immutable
 public final class BucketArgs {
```
[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/BucketArgs.java:40]: Marked `DefaultReplicationConfig`, `ECReplicationConfig`, `RatisReplicationConfig` and `StandaloneReplicationConfig`.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OzoneAcl.java [hunk 3]
```diff
@@ -49,6 +50,7 @@
  * <li>world::rw
  * </ul>
  */
+@Immutable
 public class OzoneAcl {
```
[the remark this answers is not in the document]
Comment (szetszwo) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OzoneAcl.java:54]: Failed HDDS-10745.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by xichen01** [comment]
@adoroszlai, thanks for implementing this so quickly.
So, this PR is to let the `BucketArgs`, `VolumeArgs` and `OzoneAcl` immutable, the subsequent classes, such as `OmBucketInfo` and `OmKeyArgs` will be annotated to immutable in other PR, right?

**Review by xichen01 — APPROVED** [review]

**Comment by adoroszlai** [comment]
Thanks @szetszwo, @xichen01 for the review.

**Comment by smengcl** [comment]
fyi I hit `java: package net.jcip.annotations does not exist` with this when trying to run a test inside IntelliJ, despite having recompiled the project again and again.

There might be a bug in IntelliJ IDEA 2024.1 that it is not able to resolve the transitive dependency symbols under some conditions:

Symptom

I was able to work around this with a explicit dependency declaration (and trigger [Load Maven Changes](https://www.jetbrains.com/help/idea/delegate-build-and-run-actions-to-maven.html#maven_reimport)):

```diff
diff --git a/hadoop-ozone/client/pom.xml b/hadoop-ozone/client/pom.xml
index a5a4364361..3c4f154090 100644
--- a/hadoop-ozone/client/pom.xml
+++ b/hadoop-ozone/client/pom.xml
@@ -53,6 +53,13 @@ https://maven.apache.org/xsd/maven-4.0.0.xsd">
       <type>test-jar</type>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>net.jcip</groupId>
+      <artifactId>jcip-annotations</artifactId>
+      <type>test-jar</type>
+      <scope>test</scope>
+      <version>1.0</version>
+    </dependency>
     <dependency>
       <groupId>org.apache.ozone</groupId>
       <artifactId>hdds-hadoop-dependency-test</artifactId>
```

Writing this down just in case anyone else hits this as well.

**Pull request merged**